server/aesdsocket.c

```diff
@@ -0,0 +1,393 @@
+        return FAILURE;
+    }
+    
+    if (SUCCESS != setsockopt(socket_fd, SOL_SOCKET, SO_REUSEADDR, &enable_reuse,sizeof(enable_reuse)))
+    {
+        syslog(LOG_PERROR, "setsockopt: %s", strerror(errno));
+        if (NULL != serverInfo)
+        {
+            freeaddrinfo(serverInfo);
+        }
+        close(socket_fd);
+        return FAILURE;
+    }
+    /* bind the socket to port */
+    if (SUCCESS != bind(socket_fd, serverInfo->ai_addr,
+                        serverInfo->ai_addrlen))
+    {
+        syslog(LOG_PERROR, "bind: %s", strerror(errno));
+        if (NULL != serverInfo)
+        {
+            freeaddrinfo(serverInfo);
+        }
+        close(socket_fd);
+        return FAILURE;
+    }
+    
+    /* free serverinfo after bind */
+    if (NULL != serverInfo)
+    {
+        freeaddrinfo(serverInfo);
+    }
+    
+    /* start daemon if flag is enabled */
+    if (start_as_daemon)
+    {
+        if (FAILURE == start_daemon())
+        {
+            close(socket_fd);
+            return FAILURE;
+        }
+    }
+
+    /* listen for connection on the socket */
+    if (SUCCESS != listen(socket_fd, MAX_CONNECTIONS_ALLOWED))
+    {
+        syslog(LOG_PERROR, "listen: %s", strerror(errno));
+        close(socket_fd);
+        return FAILURE;
+    }    
+    
+
+    /* exit accepting connections once signal is received */
+    while (!exit_condition)
+    {
+        /* accept the connection on the socket */
+        connection_fd = accept(socket_fd, (struct sockaddr *)&clientAddr, &clientAddrLen);
+        if (FAILURE == connection_fd)
+        {
+            syslog(LOG_PERROR, "accept: %s", strerror(errno));
+            close(socket_fd);
+            return FAILURE;
+        }  
+    
+        /* converts binary ip address to string format */
+        if (NULL == inet_ntop(AF_INET, &(clientAddr.sin_addr), ClientIpAddr, INET_ADDRSTRLEN))
+        {
+            syslog(LOG_PERROR, "inet_ntop: %s", strerror(errno));
+        }
+        syslog(LOG_INFO, "Accepted connection from %s", ClientIpAddr);
+    
+        /* open file in readwrite mode */
+        file_fd = open(FILENAME, O_CREAT|O_RDWR|O_APPEND, 
+                       S_IWUSR | S_IRUSR | S_IWGRP | S_IRGRP | S_IROTH);
+        if (FAILURE == file_fd)
+        {
+            syslog(LOG_ERR, "Error opening %s file: %s", FILENAME, strerror(errno));
+            close(connection_fd);
+            close(socket_fd);
+            return FAILURE;
+        }
+
+        /* loop to receive data until new line is found */
+        do
+        {
+            memset(buffer, 0, MAX_BUFF_LEN);
+            /* recv data from client */
+            recv_bytes = recv(connection_fd, buffer, MAX_BUFF_LEN, 0);
+            if (FAILURE == recv_bytes)
+            {
+                syslog(LOG_PERROR, "recv: %s", strerror(errno));
+                close(file_fd);
+                close(connection_fd);
+                close(socket_fd);
+                return FAILURE;
```

Comment:
@chchbuff consider a design where you can exit gracefully through one point in main rather than duplicating cleanup logic.